### ISSUE TITLE
Gateways should be able to easily override the card update form

### ISSUE THREAD

**Issue by chrisvanpatten** [restrictcontentpro/restrict-content-pro#1475]
Currently, gateway plugins would need to provide a custom override for the `card-update-form.php` template if they want to change the display of the card update form fields.

This is an issue for #1462 and other gateways that provide their own credit card forms, particularly for reducing PCI scope.

My proposal (moved here from #1462 where I originally drafted it, but moved it here because it's really a standalone issue that applies to multiple gateways):

_(Please poke holes in this, I'm sure there are many.)_

+ Gateways should be able to define `$this->supports[] = 'card-updates';` and this setting should be checked in `rcp_member_can_update_billing_card()` instead of the [current conditionals](https://github.com/restrictcontentpro/restrict-content-pro/blob/7d90a8565bfbf65690158870b5e45cb8b00de0cd/includes/member-functions.php#L1207-L1220).
+ The [code that defines the fields](https://github.com/restrictcontentpro/restrict-content-pro/blob/master/templates/card-update-form.php#L36-L69) in `card-update-form.php` would be moved to its own template (`card-update-form-fields.php` or something).
+ If the gateway supports card-updates, the section in `card-update-form.php` that _used_ to have the fields would now load `$current_gateway->update_fields()`.
+ The `update_fields()` method in the base gateway class would just load the new `card-update-form-fields.php` template, so it would be effectively identical to the current behaviour, loading the same code in the same place. Then, gateways could provide their own `update_fields()` method allowing them to override these fields at the gateway-level by loading a custom template file or inserting code inline.

**Comment by mindctrl** [comment]
@chrisvanpatten thanks for this! It's definitely something we'll need to address as we support hosted fields. I haven't given it a lot of thought yet, but this generally sounds like a good approach.

**Comment by chrisvanpatten** [comment]
Thanks @mindctrl! This is particularly pressing for Stripe, since the old Stripe method (custom form) is being deprecated in favour of Elements, but I think it'll be particularly good in enabling easier support for 2nd and 3rd tier legacy gateways who only offer `<iframe>`-based or other hosted forms. (Shockingly/frustratingly common, I'm learning!)

**Comment by chrisvanpatten** [comment]
Would love to start putting a PR together to incorporate a fix for this. Maybe this should be a first step before moving on to #1462 in a later release?

Would it make sense to post a WIP pull request and adjust based on that?

**Comment by nosegraze** [comment]
Related: I got a request from a customer who is using Stripe Checkout and wanted the billing card update form to use the Stripe Checkout modal.

**Comment by chrisvanpatten** [comment]
Hey @nosegraze @mindctrl we'd be really interested in getting Stripe Elements support in core (#1462) and merging in our plugin, but this would be a blocker to that (#1575 would benefit as well). What would be a good roadmap to tackling this? Happy to provide a PR that builds in this functionality, then a separate PR for #1462.

**Comment by mindctrl** [comment]
@chrisvanpatten Sorry about the delay here. Our priorities are a bit shifted at the moment, probably for a next couple weeks. We'd love to see a PR if you have the time.

**Comment by chrisvanpatten** [comment]
No prob, I totally understand. We'll work up a WIP PR that you can review!

**Comment by nosegraze** [comment]
@chrisvanpatten Just wondering if this is still something you're working on? Do you have any questions or is there anything we can help with? This could be a good candidate for the `3.0` release.

**Comment by chrisvanpatten** [comment]
@nosegraze Yes! Our dev handling it is out of the country right now which could impact the timeline though. I'll see if I can get back up to speed and finish up a WIP myself.

**Comment by nosegraze** [comment]
@chrisvanpatten Okay no problem - thanks for the update. :) It doesn't necessarily have to be in `3.0`, so no rush if you can't make that..

**Comment by mindctrl** [comment]
Related to this, `[rcp_update_card]` uses `rcp_member_can_update_billing_card()` to check if members can update their cards, and it checks for PayPal and Authorize.net, but neither of those gateways hook into the `rcp_get_card_details` filter inside `RCP_Member::get_card_details()` so they don't provide update card functionality.

**Comment by chrisvanpatten** [comment]
Working on a project right now that would benefit from this, so I'll see if there's a way we can get an initial PR up as part of that work.
